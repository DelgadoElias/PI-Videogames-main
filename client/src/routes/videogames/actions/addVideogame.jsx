import axios from "axios";
import { useState } from "react";
import { useHistory } from "react-router";
import { useSelector } from "react-redux";
import vStyles from "../../../assets/styles/add.module.css";

import { Link } from "react-router-dom";
import MainLayout from "../../../components/Layouts/mainLayout";

/**
 * AddVideogame route - Renders the videogame form route to add new videogames
 * @return {React.ReactNode}
 */
export default function AddVideogame() {
  const genres = useSelector((state) => state.genres);
  const platforms = useSelector((state) => state.platforms);
  const history = useHistory();
  const [videogame, setVideogame] = useState({
    name: "",
    description: "",
    image: "",
    rating: 0,
    released: "",
    genres: [],
    platforms: [],
  });

  /**
   * onInputChange - Handle input changes of the videogame properties
   * @param {Event} e - HTML Events to fire e.preventDefault()
   * @description preventDefault function and set the videogame properties to the state
   */
  function onInputChange(e) {
    e.preventDefault();
    setVideogame({
      ...videogame,
      [e.target.name]: e.target.value,
    });
  }

  // ..... Cambios del checkbox nomás .....
  function handleCheckbox(e) {
    if (e.target.checked) {
      // Cuándo esté checkeado, subimelo
      setVideogame({
        ...videogame,
        genres: [e.target.value, ...videogame.genres],
      });
    } else {
      let filteredGenres = videogame.genres.filter((x) => {
        return x !== e.target.value;
      });
      setVideogame({
        ...videogame,
        genres: [...filteredGenres],
      });
    }
  }

  // ..... Cambios del selector .....
  function onSelect(e) {
    setVideogame({
      ...videogame,
      platforms: [...videogame.platforms, e.target.value],
    });
  }
  // Eliminar las platforms cuando nos arrepentimos
  function onDeSelect(e) {
    let filtraciones = videogame.platforms.filter((x) => {
      return x !== e;
    });

    setVideogame({
      ...videogame,
      platforms: [...filtraciones],
    });
  }

  async function onSubmit(e) {
    e.preventDefault();

    axios
      .post("http://localhost:3001/videogame/", videogame)
      .then(() => {
        history.push("/");
      })
      .catch((err) => {
        console.log(err);
      });
  }

  return (
    <MainLayout>
      <div
        className={`${vStyles.container} ${vStyles.CardItems} ${vStyles.animated} ${vStyles.fadeIn} ${vStyles.fast}`}
      >
        {/* ..... Button ..... */}
        <Link to="/home">
          <button className={`${vStyles.input} ${vStyles.button}`}>
            <h3>Back to Home</h3>
          </button>
        </Link>

        {/* ..... Required form ..... */}
        <form onSubmit={onSubmit}>
          <div>
            <div>
              <h1 styles="text-align: left">¡Add new videogames!</h1>
              <br />
            </div>

            {/* ..... Name ..... */}
            <div>
              <label>Nombre:</label>
              <br />
              <input
                className={vStyles.input}
                onChange={onInputChange}
                name="name"
                type="text"
                value={videogame.name}
                required
              />
            </div>

            {/* ..... Description ..... */}
            <div>
              <label>Descripción:</label>
              <br />

              <input
                className={vStyles.input}
                onChange={onInputChange}
                name="description"
                type="text"
                value={videogame.description}
                required
              />
            </div>

            {/* ..... Image URL ..... */}
            <div>
              <label>Imágen:</label>
              <br />

              <input
                className={vStyles.input}
                onChange={onInputChange}
                name="image"
                type="text"
                value={videogame.image}
              />
            </div>

            {/* ..... Rating ..... */}
            <div>
              <label>Rating:</label>
              <br />

              <input
                className={vStyles.input}
                onChange={onInputChange}
                name="rating"
                type="number"
                value={videogame.rating}
              />
            </div>

            {/* ..... Released ..... */}
            <div>
              <label>Lanzamiento:</label>
              <br />

              <input
                className={vStyles.input}
                onChange={onInputChange}
                name="released"
                type="text"
                value={videogame.released}
              />
            </div>

            {/* ..... Géneros ..... */}
            <h3>Géneros</h3>
            <div className={vStyles.father}>
              <br />
              <br />

              {/* ..... Checkbox para los géneros seleccionados ..... */}
              {genres?.map((x) => {
                let genreSplited = x.name.split(" ");
                return (
                  <label>
                    <input
                      className={vStyles.hijo}
                      onChange={(e) => handleCheckbox(e)}
                      type="checkbox"
                      name="genres"
                      value={x.name}
                    ></input>
                    {genreSplited[1] ? "MMO" : genreSplited[0]}
                  </label>
                );
              })}
            </div>

            {/* ..... Plataformas ..... */}

            <div className={vStyles.containerMini}>
              <h3>Platforms</h3>
              <br />

              <select className={vStyles.input} onChange={(e) => onSelect(e)}>
                {platforms?.map((x) => {
                  return (
                    <option name="platforms" value={x.name}>
                      {x.name}
                    </option>
                  );
                })}
              </select>

              <div>
                {/* ..... Plataformas Seleccionadas ..... */}
                <h3>Platforms selected:</h3>
                <ul>
                  {videogame.platforms?.map((x) => {
                    return (
                      <h4 value={x} onClick={(e) => onDeSelect(x)}>
                        {x} - <span className={vStyles.x}>X</span>
                      </h4>
                    );
                  })}
                </ul>
              </div>
            </div>

            <hr />
            {/* Botón de envío */}
            <input
              className={`${vStyles.input} ${vStyles.padder}`}
              type="submit"
            />
            <br />
          </div>
        </form>
      </div>
    </MainLayout>
  );
}
