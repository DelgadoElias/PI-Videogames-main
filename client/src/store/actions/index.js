
import axios from 'axios'

// ----------------------------------------------------------------
// ..
export const FETCH_VIDEOGAMES = 'FETCH_VIDEOGAMES'
export const FETCH_GENRES = 'FETCH_GENRES'
export const FETCH_PLATFORMS = 'FETCH_PLATFORMS'

// ..
export const SEARCH_VIDEOGAMES = 'SEARCH_VIDEOGAMES'

// ..
export const SORT_VIDEOGAMES = 'SORT_VIDEOGAMES'
export const SORT_RATING = 'SORT_RATING'

// ..
export const DB_FILTER = 'DB_FILTER'
export const API_FILTER = 'API_FILTER'
export const GENRES_FILTER = 'GENRES_FILTER'
export const PLATFORMS_FILTER = 'PLATFORMS_FILTER'


// ..
export const POST_VIDEOGAME = 'POST_VIDEOGAME'

// ----------------------------------------------------------------
export const GAME_RANDOM = 'GAME_RANDOM'
export const RESET_GAMES = 'RESET_GAMES'


// ----------------------------------------------------------------

// Fetch/gets ..... ..... ..... ..... .....

// Videogames
export function fetchVideogames(){
    return async function(dispatch){
        try {
            let poder = await axios.get('http://localhost:3009/videogames') // Hacemos al backEnd nuestro
            dispatch({
            type: FETCH_VIDEOGAMES,
            payload: poder.data
            })
        } catch (e) {
            console.log(e)
        }
        
    }
}

// Géneros
export function fetchGenres(){
    return async function(dispatch) {
        try {
            let generos = await axios.get('http://localhost:3009/genres');
            dispatch({
            type: FETCH_GENRES,
            payload: generos.data
            })
        } catch (e) {
            console.log(e);
        }
    }
}

// Plataformas
export function fetchPlatforms(){
    return async function(dispatch) {
        try {
            let plataformas = await axios.get('http://localhost:3009/platforms');
            dispatch({
            type: FETCH_PLATFORMS,
            payload: plataformas.data
            })
        } catch (e) {
            console.log(e);
        }
    }
}


// Searches ..... ..... ..... ..... ..... .....

export function searchVideogames(search, page = 1){
    return async function(dispatch){
        try {
            let poder = await axios.get(`http://localhost:3009/videogames?name=${search}&page=${page}`) // Hacemos al backEnd nuestro
            dispatch({
            type: SEARCH_VIDEOGAMES,
            payload: poder.data
            })
        } catch (e) {
            console.log(e)
        }
        
    }
}

// Ordenamientos ..... ..... ..... ..... ..... ..... .....

// Alfabético
export function sort(order){
    return {
        type : SORT_VIDEOGAMES,
        payload : order
        
    }
}

// Rating ----- ----- ----- -----
export function sortRating(order){
    return {
        type : SORT_RATING,
        payload : order
        
    }
}


// Filtering ..... ..... ..... ..... ..... ..... ..... ....
// Only Db
export function dbFilter(type){
    return {
        type : DB_FILTER,
        payload : type
    }
}

// By genres.
export function genresFilter(type){
    return {
        type : GENRES_FILTER,
        payload : type
    }
}

// By platforms.
export function platformsFilter(type){
    return {
        type : PLATFORMS_FILTER,
        payload : type
    }
}

// By API
export function apiFilter(type){
    return {
        type: API_FILTER,
        payload : type
    }
}

// Extras ..... ..... ..... ..... ..... ..... ..... ..... 

// Pick a random game and render in the page
export function gameRandom(){
    return {
        type: GAME_RANDOM,
        payload: "none"
    }
}

// Reset videogame cards

export function resetGames(){
    return {
        type: RESET_GAMES,
        payload: "none"
    }
}