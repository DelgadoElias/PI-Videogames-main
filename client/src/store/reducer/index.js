import { NONE } from "../../constantes/filters";
import { ASCENDENTE } from "../../constantes/sort";
import {
  API_FILTER,
  DB_FILTER,
  FETCH_GENRES,
  FETCH_PLATFORMS,
  FETCH_VIDEOGAMES,
  GAME_RANDOM,
  GENRES_FILTER,
  PLATFORMS_FILTER,
  resetGames,
  RESET_GAMES,
  SEARCH_VIDEOGAMES,
  SORT_RATING,
  SORT_VIDEOGAMES,
} from "../actions";

const initialState = {
  videogames: [],
  filteredVideogames: [],
  platforms: [],
  genres: [],
  filteredGenres: [],
  filteredPlatforms: [],
  backup: [],
  pickEm: {},
};


// No es buena idea duplicar data front a back

export default function reducer(state = initialState, action) {
  switch (action.type) {
    // **************************************************************************************************
    // -------------------------------------
    case FETCH_VIDEOGAMES:
      return {
        ...state,
        videogames: action.payload,
        filteredVideogames: action.payload,
        backup : [...state.filteredVideogames]
      };
    // -------------------------------------
      case FETCH_GENRES:
        return {
          ...state,
          genres: action.payload,
          filteredGenres: action.payload
        };

    // -------------------------------------
    case FETCH_PLATFORMS:
      return {
        ...state,
        platforms: action.payload,
        filteredPlatforms: action.payload
      };

    // **************************************************************************************************
    // -------------------------------------
    case SEARCH_VIDEOGAMES:
      if(typeof action.payload === 'string'){
        console.log('Error: ' + action.payload);
        return {
          ...state,
        }
      }else{
        return {
          ...state,
          filteredVideogames: action.payload,
        };
      }

    // **************************************************************************************************
    // -------------------------------------
    case SORT_VIDEOGAMES:
      // Optional: Cambiarlo a función como hicimos en el M1
      let orderedVideogames = state.filteredVideogames.sort((a, b) => {
        if (a.name < b.name) return action.payload === ASCENDENTE ? -1 : 1;

        if (a.name > b.name) return action.payload === ASCENDENTE ? 1 : -1;
        return 0;
      });
      return {
        ...state,
        filteredVideogames: [...orderedVideogames],
      };
    // -------------------------------------
    case SORT_RATING:
      let orderedRatings = [...state.videogames];

      // Este lo dejamos así nomás, demostramos usar todos los métodos
      orderedRatings = state.filteredVideogames.sort((a, b) => {
        if (a.rating < b.rating) return action.payload === ASCENDENTE ? -1 : 1;

        if (a.rating > b.rating) return action.payload === ASCENDENTE ? 1 : -1;
        return 0;
      });
      return {
        ...state,
        filteredVideogames: [...orderedRatings],
      };

    // **************************************************************************************************
    // -------------------------------------
      case DB_FILTER:
        
        if(action.payload === true){
          let filteredItems = state.videogames.filter((x) => {
            // Miremos la propiedad especial
            return x.hasOwnProperty('createdInDb')}
          );
          // ..... ..... ..... .....
          return {
            ...state,
            backup : [...state.filteredVideogames],
            filteredVideogames : [...filteredItems]
          }
        }else{ 
          // ..... ..... ..... .....
          return { // Si no es lo volvemos a setear
            ...state,
            filteredVideogames : [...state.backup]
          }
        }

    // -------------------------------------
    case API_FILTER:
        if(action.payload === true){
          let filteredItems = state.videogames.filter((x) => {
            // Miremos el tamaño del id
            let superId = x.id.toString();
            return superId.length < 8 }
          );
          // ..... ..... ..... .....
          return {
            ...state,
            backup : [...state.filteredVideogames],
            filteredVideogames : [...filteredItems]
          }
        }else{ 
          // ..... ..... ..... .....
          return { // Si no es lo volvemos a setear
            ...state,
            filteredVideogames : [...state.backup]
          }
        }    

    // -------------------------------------
        case GENRES_FILTER:

          if(action.payload === NONE){
            return {
              ...state,
              filteredVideogames : [...state.videogames]
            }
          }else{ // Caso contrario, tenemos un género
            let filterByGenre = state.videogames.filter((x) => {
              
              // Revisamos el array.
              for(let i = 0; i < x.genres.length; i++) {
                if(x.genres[i].name === action.payload){
                  return true;
                }
              }
              // Si no lo encontramos en el array no hay nada
              return false; 
              
            });
            // ..... Finalización filter .....
            return {
              ...state,
              filteredVideogames: [...filterByGenre]
            }
          }
          
        // -------------------------------------

          case PLATFORMS_FILTER:

          if(action.payload === NONE){
            return {
              ...state,
              filteredVideogames : [...state.videogames]
            }
          }else{ 
            let filterByPlatform = state.videogames.filter((x) => {

              // Revisamos el array.
              for(let i = 0; i < x.platforms.length; i++) {
                let consta = x.platforms[i].platform ? x.platforms[i].platform.name : x.platforms[i].name 
                if(consta === action.payload){
                  return true;
                }
              }
              return false; 
              
            });

            // ..... Finalización filter .....
            return {
              ...state,
              filteredVideogames: [...filterByPlatform]
            }
          }

    // **************************************************************************************************
    // -------------------------------------
          case GAME_RANDOM:
            
            let one = Math.floor(Math.random() * (state.videogames.length ));
          return {
            ...state,
            pickEm: state.videogames[one]
          }
    // -------------------------------------

          case RESET_GAMES:
            return {
              ...state,
              filteredVideogames: [...state.videogames]
            }
    // -------------------------------------
    
    default:
      return state;
  }
}
