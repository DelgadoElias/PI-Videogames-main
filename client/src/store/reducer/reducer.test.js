import reducer from './index'
import {initialState} from './index'

// Acciones
import {
    DB_FILTER,
    FETCH_GENRES,
    FETCH_PLATFORMS,
    FETCH_VIDEOGAMES,
    GENRES_FILTER,
    PLATFORMS_FILTER,
    SEARCH_VIDEOGAMES,
    SORT_RATING,
    SORT_VIDEOGAMES,
  } from "../actions";

  import {fetchVideogames} from "../actions"

  

describe('reducer', () => {
    // Initial State ?
    it('Debería devolver el estado inicial', () => {
      expect(reducer(undefined, {})).toMatchSnapshot()
    })

    it('Debería traer los videojuegos cuando llamamos a fetchVideogames', () => {
        // fijarse bien las propiedades que tiene que recibir.
        expect(reducer(initialState, fetchVideogames())).toHaveProperty('videogames',[]);
        console.log(reducer(initialState, fetchVideogames()));

    });

  })