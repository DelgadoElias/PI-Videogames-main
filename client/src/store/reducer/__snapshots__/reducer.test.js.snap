// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer Deber√≠a devolver el estado inicial 1`] = `
Object {
  "backup": Array [],
  "filteredGenres": Array [],
  "filteredPlatforms": Array [],
  "filteredVideogames": Array [],
  "genres": Array [],
  "platforms": Array [],
  "videogames": Array [],
}
`;

exports[`reducer should return the initial state 1`] = `
Object {
  "backup": Array [],
  "filteredGenres": Array [],
  "filteredPlatforms": Array [],
  "filteredVideogames": Array [],
  "genres": Array [],
  "platforms": Array [],
  "videogames": Array [],
}
`;
